File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.py
def solution(book_time):
    room = []
    
    for start, end in sorted(book_time):
        start_h, start_m = map(int, start.split(":"))
        end_h, end_m = map(int, end.split(":"))
        start_time = start_h * 60 + start_m
        end_time = end_h * 60 + end_m + 10
        
        room.sort()
        if len(room) == 0 or room[0] > start_time:
            room.append(end_time)
        else:
            room[0] = end_time
    return len(room)

File: 프로그래머스/2/155651. 호텔 대실/호텔 대실.java
import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;
import java.util.PriorityQueue;


class Solution {
    public int solution(String[][] book_time) {
        int answer = 0;
        
        Arrays.sort(book_time, (o1, o2) -> {
            if (o1[0].equals(o2[0])) return o1[1].compareTo(o2[1]);
            else return o1[0].compareTo(o2[0]);
        });
        
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        for (String[] time: book_time){
            int start = convertTimeStr2Int(time[0]);
            int end = convertTimeStr2Int(time[1]) + 10;
            
            if (!pq.isEmpty() && pq.peek() <= start){
                pq.poll();
            }
            pq.offer(end);
            answer = Math.max(answer, pq.size());
        }
        
        return answer;
    }
    
    public int convertTimeStr2Int(String strTime){
        return Integer.parseInt(strTime.split(":")[0]) * 60 + Integer.parseInt(strTime.split(":")[1]);
    }
}